add a from method to ContractClient to instantiate from a contractId

BlaineHeffron opened the pull request:
Adds a `from` method which uses the `contractId` to create an instance of the `ContractClient` by retrieving the wasm from the blockchain and extracting its `ContractSpec`. I also add a `fromWasm` method which bypasses the wasm retrieval portion if you already have the wasm buffer.

willemneal commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,45 @@ export class ContractClient {
+    if (!options || !options.rpcUrl || !options.contractId) {
+      return Promise.reject(new TypeError('options must contain rpcUrl and contractId'));
+    }
+    const { rpcUrl, contractId, allowHttp } = options;
```
Could you move this destructuring to function signature?

willemneal commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,45 @@ export class ContractClient {
+    const serverOpts: Server.Options = { allowHttp: allowHttp?? !rpcUrl.startsWith('https') };
+    const server = new Server(rpcUrl, serverOpts);
+    const contractLedgerKey = new Contract(contractId).getFootprint();
+    const response = await server.getLedgerEntries(contractLedgerKey);
```
Perhaps the contract --> wasm lookup logic could be a method offered by `server`. Seems useful outside of this usecase.

Shaptic replied:
Agreed. It's [documented](https://developers.stellar.org/network/soroban-rpc/api-reference/methods/getLedgerEntries#requesting-a-contracts-wasm-code) but a built-in would be better; I opened https://github.com/stellar/js-stellar-sdk/issues/961.

BlaineHeffron replied:
added, will add unit tests as well

Shaptic commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,45 @@ export class ContractClient {
+    const response = await server.getLedgerEntries(contractLedgerKey);
+    if (!response.entries[0]?.val)
+      return Promise.reject(new Error(`Could not obtain contract from server`));
+    const wasmHash = ((response.entries[0].val.value() as xdr.ContractDataEntry).val().value() as xdr.ScContractInstance).executable().wasmHash();
```
You should use the type-specific fetch methods to avoid casting here, see e.g. curr.d.ts:

```suggestion
    const wasmHash = response.entries[0].val.contractData().val().contractInstance().executable().wasmHash();
```

BlaineHeffron replied:
thanks for the tip, fixed

Shaptic commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,45 @@ export class ContractClient {
+      hash: wasmHash,
+    }));
+    const responseWasm = await server.getLedgerEntries(ledgerKeyWasmHash);
+    const wasmBuffer = (responseWasm.entries[0].val.value() as xdr.ContractCodeEntry).code();
```
Same here: 
```suggestion
    const wasmBuffer = responseWasm.entries[0].val.contractCode().code();
```

Shaptic commented on src/contract_client/utils.ts:
```diff
@@ -1,7 +1,11 @@
  * The default timeout for waiting for a transaction to be included in a block.
  */
-export const DEFAULT_TIMEOUT = 5 * 60;
+export const DEFAULT_TIMEOUT = 10;
```
Intentionally undoing? Looks like a bad upstream sync/merge.

BlaineHeffron replied:
Good catch, sorry about that!

Shaptic commented on src/contract_client/utils.ts:
```diff
@@ -1,7 +1,11 @@
+// @ts-ignore
+import * as jsxdr from "@stellar/js-xdr";
```
Import only what you need!

```suggestion
import { XdrReader } from "@stellar/js-xdr";
```

Shaptic commented on src/contract_client/utils.ts:
```diff
@@ -1,7 +1,11 @@
+// @ts-ignore
+import * as jsxdr from "@stellar/js-xdr";
+import { xdr } from "@stellar/stellar-base";
```
Using your own instance is better, I think?

```suggestion
import { xdr } from "..";
```

Shaptic commented on test/e2e/initialize.sh:
```diff
@@ -22,7 +22,7 @@ else
 fi
 
-NETWORK_STATUS=$(curl -s -X POST "$SOROBAN_RPC_URL" -H "Content-Type: application/json" -d '{ "jsonrpc": "2.0", "id": 8675309, "method": "getHealth" }' | sed 's/.*"status":"\(.*\)".*/\1/')
+NETWORK_STATUS=$(curl -s -X POST "$SOROBAN_RPC_URL" -H "Content-Type: application/json" -d '{ "jsonrpc": "2.0", "id": 8675309, "method": "getHealth" }' | sed -n 's/.*"status":\s*"\([^"]*\)".*/\1/p')
```
should we just use `jq` :joy:

willemneal replied:
Would be more readable.

chadoh replied:
Pros: way more readable
Cons: requires people to install a separate shell dependency, in order to use this script

¯\\\_(ツ)_/¯

Shaptic replied:
hmm, good point @chadoh! I think we should evangelize `jq` but that's probably too opinionated :thinking: 

eh honestly people are installing thousands of packages when they `npm i`.... what's one more?

BlaineHeffron replied:
Oh sorry I missed Chad's comment. That was my feeling about it which is why I just used sed. I'll just keep it with sed I suppose?

Shaptic commented on src/soroban/server.ts:
```diff
@@ -235,6 +235,82 @@ export class Server {
+   *   console.error("Error fetching WASM bytecode:", err);
+   * });
+   */
+  public async getContractWasm(
```
yoooooooooooooo :fire: :fire: :fire: good stuff :clap:

Shaptic replied:
but now you need to add tests to `/test/unit/server/soroban` :p 

if you don't want to (totally fine, it's out of scope here), you can just make this a private helper, instead, and I'll do the piping to add it to `Server` as part of #961.

BlaineHeffron replied:
No problem Ill add a test.

BlaineHeffron replied:
Well that took longer than I expected but it was a good learning experience!

Shaptic replied:
I warned ya ;) but thank you! The tests look thorough, great work :clap:

Shaptic reviewed (commented):
:clap:

willemneal commented on test/e2e/src/util.js:
```diff
@@ -91,3 +91,22 @@ async function clientFor(contract, { keypair = generateFundedKeypair(), contract
+    publicKey,
+    ...wallet,
+  };
+  const client = await ContractClient.from(options);
```
```suggestion
  return await ContractClient.from(options);
```
```suggestion
  const client = await ContractClient.from(options);
```

willemneal commented on test/e2e/src/util.js:
```diff
@@ -91,3 +91,22 @@ async function clientFor(contract, { keypair = generateFundedKeypair(), contract
+    ...wallet,
+  };
+  const client = await ContractClient.from(options);
+  return client;
```
```suggestion
```

chadoh commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,35 @@ export class ContractClient {
+  /**
+   * Generate a ContractClient instance from the ContractClientOptions and the wasm binary
+   */
+  static async fromWasm(options: ContractClientOptions, wasm: BufferSource): Promise<ContractClient> {
```
nit: the constructor takes `spec` then `options`, whereas this takes `options` then `wasm`. I'd prefer `options` always came in the same order. Second, probably? To avoid the breaking change with making `constructor` take it as the first arg.

chadoh commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,35 @@ export class ContractClient {
+    const wasmModule = await WebAssembly.compile(wasm);
+    const xdrSections = WebAssembly.Module.customSections(wasmModule, "contractspecv0");
+    if (xdrSections.length === 0) {
+      return Promise.reject(new Error('Could not obtain contract spec from wasm'));
```
I haven't seen this "in case of error within `async` function, return a `Promise.reject`" pattern before. Is this codebase doing this elsewhere? I've most commonly just seen a `throw new Error` instead. Which also seems to be what this StackOverflow answer recommends: https://stackoverflow.com/a/42453705

chadoh replied:
@Shaptic do you have opinions on `return Promise.reject(new Error(...))` vs `throw new Error(...)`?

BlaineHeffron replied:
changed to throw Error instead

Shaptic replied:
no opinions tbh, but throw seems more readable to a non-`Promise`r

BlaineHeffron replied:
So I ended up going back to promise.reject because it was convenient for that test case in the mocha test file which I patterned after the `get_contract_data_test.js`. I noticed that server.ts was returning a promise.reject when it wasn't getting entries back from the server that it expected, so it was after an await call whereas it was throwing errors for stuff like malformed user input (TypeError for example). So based on that pattern it made sense for me to return a promise.reject with error code 404.

chadoh commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,35 @@ export class ContractClient {
+   */
+  static async from(options: ContractClientOptions): Promise<ContractClient> {
+    if (!options || !options.rpcUrl || !options.contractId) {
+      return Promise.reject(new TypeError('options must contain rpcUrl and contractId'));
```
Same thing here

chadoh commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,35 @@ export class ContractClient {
+      return Promise.reject(new TypeError('options must contain rpcUrl and contractId'));
+    }
+    const { rpcUrl, contractId, allowHttp } = options;
+    const serverOpts: Server.Options = { allowHttp: allowHttp?? !rpcUrl.startsWith('https') };
```
`options` [has an `allowHttp` field](https://github.com/AhaLabs/js-stellar-sdk/blob/3f134730d015602bf8f964a3c7334c8fb372bff7/src/contract_client/types.ts#L76); we should prefer that instead of inferring it from the `rpcUrl`

BlaineHeffron replied:
fixed

chadoh commented on test/e2e/src/util.js:
```diff
@@ -91,3 +91,22 @@ async function clientFor(contract, { keypair = generateFundedKeypair(), contract
+/**
+ * Generates a ContractClient given the contractId using the from method.
+ */
+async function clientForFromTest(contractId, publicKey, keypair) {
```
I'm in favor of replacing the guts of `clientFor` with the new `ContractClient.from` logic. Then you can skip adding any special tests for `from`, since they'll all use it. `from` already uses the constructor internally, so we're implicitly testing everything already, and then we avoid shelling out to the CLI from within `clientFor`

chadoh replied:
Correction: we could get rid of the `xdr` lines inside the `contracts` definition. We could keep shelling out for the `contract deploy` logic, for now.

chadoh replied:
If we do this, then we probably want to add one special test for the reverse: that a spec generated with the CLI can still work with `ContractClient`. We probably want to test it explicitly here, since the TS Bindings in the CLI will still want to use that, to avoid fetching the XDR redundantly in everyone's browser at runtime.

BlaineHeffron replied:
Hmm, looking into this since we already have the wasmHash we should really allow the user to get the contract spec from their wasmHash since they might have that from producing it with the cli (similar to how I have it so you can get it from their wasm). So I could split the `Server.getContractWasm` into two methods: getContractWasmFromId and getContractWasmFromWasmHash or something like that. Then client.ts could use the latter if the user supplies it (getFromWasmHash). Thoughts?

BlaineHeffron replied:
> If we do this, then we probably want to add one special test for the reverse: that a spec generated with the CLI can still work with `ContractClient`. We probably want to test it explicitly here, since the TS Bindings in the CLI will still want to use that, to avoid fetching the XDR redundantly in everyone's browser at runtime.

Makes sense

BlaineHeffron replied:
I changed it to using fromWasmHash and added one check which uses the constructor.

chadoh commented on test/e2e/src/test-custom-types.js:
```diff
@@ -144,6 +144,16 @@ test('u128', async t => {
 
+test('from', async (t) => {
+  function flattenInstance(object) {
+    return JSON.parse(JSON.stringify(object));
```
In what way does this logic flatten an object? Removes dynamic fields?

chadoh replied:
("flatten" to me means that it would turn things like `[ [1], [2, 3], 4]` into `[1, 2, 3, 4]`, but this doesn't do that)

BlaineHeffron replied:
Its not really flattening it but it works around the fact that deepEquals requires the constructors to be the same, see https://github.com/avajs/ava/issues/1481#issuecomment-516502948

BlaineHeffron replied:
Although I'm not 100% sure that it gives the exact equality we need.. which is why I also added the deepEquals for the ScSpecEntrys

BlaineHeffron replied:
Marking this resolved for now, I updated the name to reflect what its actually doing.

chadoh commented on test/e2e/src/test-custom-types.js:
```diff
@@ -144,6 +144,16 @@ test('u128', async t => {
+  }
+
+  const clientFromFrom = await clientForFromTest(t.context.contractId, t.context.publicKey, t.context.keypair);
+  t.deepEqual(flattenInstance(clientFromFrom), flattenInstance(t.context.client));
```
We do we need to "flatten" them for them to be deep equal?

chadoh commented:
Still marked as draft. Is that intended, @BlaineHeffron? Seems like it's about ready to go!

BlaineHeffron commented:
> Still marked as draft. Is that intended, @BlaineHeffron? Seems like it's about ready to go!

See my comment here - https://github.com/stellar/js-stellar-sdk/pull/960#discussion_r1595553929

Shaptic commented on src/soroban/server.ts:
```diff
@@ -235,6 +235,62 @@ export class Server {
+    const contractLedgerKey = new Contract(contractId).getFootprint();
+    const response = await this.getLedgerEntries(contractLedgerKey);
+    if (!response.entries.length || !response.entries[0]?.val) {
+      return Promise.reject({code: 404, message: `Could not obtain contract hash from server`});
```
throwing is fine here!

BlaineHeffron replied:
I changed it back to Promise.reject since I was following the pattern in the get_contract_data.js test which checks for a 404 code. I can revert it to throwing an error if you'd prefer.

Shaptic reviewed (commented):
The RPC work looks good to me, I'll wait for approval from @chadoh / @willemneal on the actual client-side work as I see there's open discussion there before merging!

BlaineHeffron commented:
Should be all ready to go now.

chadoh commented on test/e2e/src/test-custom-types.js:
```diff
@@ -1,20 +1,26 @@
 })
 
+test('hello from constructor', async t => {
+  const { client } = await clientFromConstructor('customTypes')
```
Given that this is the only test that uses `clientFromConstructor`, and might _always be_ the only such test, probably not worth extracting the logic to a function in `util`.

chadoh commented on test/e2e/src/test-custom-types.js:
```diff
@@ -144,6 +150,17 @@ test('u128', async t => {
+    return JSON.parse(JSON.stringify(object));
+  }
+
+  const clientFromFrom = await clientForFromTest(t.context.contractId, t.context.publicKey, t.context.keypair);
```
I think it might be worth moving these checks into their own file, rather than putting them in `test-custom-types.js`. Then you don't need to worry so much about having beefy helper functions in that file.

BlaineHeffron replied:
Okay, I will consolidate that logic into its own file for just testing the `from` method and the Constructor.

chadoh commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,49 @@ export class ContractClient {
+  /**
+   * Generate a ContractClient instance from the ContractClientOptions and the wasm hash 
+   */
+  static async fromWasmHash(wasmHash: Buffer, options: ContractClientOptions): Promise<ContractClient> {
```
Why take `wasmHash` a Buffer? Most of my experience working with wasm hashes comes from using the CLI, where I'm used to copy-pasting a string. I see even in the test here, you end up with a hex string that you then need to encode as a Buffer. Can we skip that, and pass a string here?

BlaineHeffron replied:
Sure, we can pass a string, perhaps with an encoding option that defaults to hex?

chadoh commented on src/soroban/server.ts:
```diff
@@ -235,6 +235,93 @@ export class Server {
+    }
+    const wasmBuffer = responseWasm.entries[0].val.contractCode().code();
+
+    return Buffer.from(wasmBuffer);
```
On that note, why encode as a Buffer here? The less we pass around Buffers, the easier it will be to remove that dependency someday. I think a "JIT conversion to Buffer" strategy is probably the best call, rather than eagerly converting to Buffer and sprinkling Buffer arguments throughout our codebase.

BlaineHeffron replied:
Good catch, no need there.

BlaineHeffron commented:
Latest commit fixes all of Chad's comments.

chadoh commented:
Tests are failing

BlaineHeffron commented:
> Tests are failing

Looks like that weird spawnSync issue we were seeing before. I added a fallback in case it doesnt work the first time.

chadoh reviewed (approved):
Thanks for all the work here! Looks great

chadoh reviewed (approved):
I think my last approval was for a single commit?

chadoh commented:
ok, I guess I just don't have the correct permissions to get this thing marked as mergeable

Shaptic commented on test/e2e/src/test-contract-client-constructor.js:
```diff
@@ -0,0 +1,110 @@
+const {
+  ContractClient,
+  basicNodeSigner,
+} = require('../../../lib/contract_client')
```
If you note the tests in the non-e2e directories, they actually pull from `StellarSdk` just like a downstream app would, where StellarSdk gets injected into `global`/`window` (see `test-nodejs.js` and` test-browser.js`). This directory uses a different framework, but we similarly want to have both browser and Node tests, but I'm not sure how to accomplish that since I didn't write these tests (cc @chadoh, would you know?).

Maybe this is a reflection of why we didn't catch the `ScVal` object thing if the other tests are running in the browser, either?

chadoh replied:
I didn't know the other tests were running in node and browser! Very cool. No, these e2e tests only run in node. Would be very cool to have them run in the browser, too. I [noted in the original ContractClient PR](https://github.com/stellar/js-stellar-sdk/pull/929) that the tests were technical debt. We should probably make an issue to track this, and make sure that we also make them work in the browser.

For now, this test is following the convention of all the other e2e tests, so maybe we leave as-is and then fix it up to match #962 once that gets merged, and finally fix them all up to run in the browser too (and maybe use mocha, like the other tests, but TBD?) as part of that follow-up.

BlaineHeffron replied:
I can leave the tests as is for now as Chad suggest and work on the e2e tests later on (its on my todo list). I think we just need to decide on if we want to convert them all to mocha or just keep using ava and I guess add browser versions of the test?

Shaptic replied:
Agreed, let's punt it :+1: imo browser versions >>> same framework but I guess both would be ideal

Shaptic reviewed (commented):
another pass :eyes: 

also be sure to run `yarn fmt`

chadoh commented on src/contract_client/utils.ts:
```diff
@@ -1,3 +1,7 @@
+// @ts-ignore
```
I think `@ts-ignore` will just ignore the next line, though I'm still not sure why that's necessary for this line.

```ts
import { XdrReader } from "@stellar/js-xdr";
```

willemneal replied:
It's not in the types file.

BlaineHeffron replied:
This is no longer an issue now that we are using cereal namespace from stellar-base

chadoh commented on src/contract_client/utils.ts:
```diff
@@ -77,8 +81,18 @@ export const contractErrorPattern = /Error\(Contract, #(\d+)\)/;
-  obj: unknown,
-): obj is { toString(): string } {
+export function implementsToString(obj: unknown): obj is { toString(): string } {
   return typeof obj === "object" && obj !== null && "toString" in obj;
```
You need the `obj !== null` because [typeof null === "object"](https://stackoverflow.com/questions/18808226/why-is-typeof-null-object) for hysterical reasons

But yeah, probably better to check that `typeof toString === "function"`

Though I think these lines were just changed because of eslint, so I'm not sure it's worth changing the implementation as part of this PR

BlaineHeffron commented on src/contract_client/utils.ts:
```diff
@@ -77,8 +81,18 @@ export const contractErrorPattern = /Error\(Contract, #(\d+)\)/;
+ */
+export function processSpecEntryStream(buffer: Buffer) {
+  const reader = new XdrReader(buffer);
+  const res: xdr.ScSpecEntry[] = [];
+  while (reader._index < reader._length) {
+    res.push(xdr.ScSpecEntry.read(reader));
+  }
+  return res;
```
eof works great!

BlaineHeffron commented on src/contract_client/utils.ts:
```diff
@@ -77,8 +81,18 @@ export const contractErrorPattern = /Error\(Contract, #(\d+)\)/;
+export function processSpecEntryStream(buffer: Buffer) {
+  const reader = new XdrReader(buffer);
+  const res: xdr.ScSpecEntry[] = [];
+  while (reader._index < reader._length) {
```
thanks

BlaineHeffron commented on src/soroban/server.ts:
```diff
@@ -235,6 +235,106 @@ export class Server {
+   *
+   * @returns {Promise<Buffer>}   a Buffer containing the WASM bytecode
+   *
+   * @throws {Error} If the contract or its associated WASM bytecode cannot be
+   * :
+   *    found on the network.
```
Good catch.

BlaineHeffron commented on src/contract_client/client.ts:
```diff
@@ -45,6 +47,68 @@ export class ContractClient {
+    options: ContractClientOptions, 
+    format: "hex" | "base64" = "hex"
+  ): Promise<ContractClient> {
+    if (!options || !options.rpcUrl ) {
```
Not sure why it failed linter check.

BlaineHeffron commented:
I addressed comments from @Shaptic although this now is dependent on https://github.com/stellar/js-stellar-base/pull/744 being merged first. The latest changes remove the external js-xdr dependency and uses the `ceareal` namespace from `stellar-base` for `XdrReader`.

Shaptic reviewed (approved):
:heart:
